File: problem_032.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "swiftc problem_0032.swift -o problem_0032" -*-
// Copyright (c) 2016 Michael Caldwell

print("Project Euler - Problem 32:")
print("Find the sum of all numbers that can be written as pandigital products.\n")

func isPandigital(var a:Int, var b:Int, var c:Int) -> Bool {
    var count = [1, 0, 0, 0, 0, 0, 0, 0, 0, 0]

    while a > 0 {
        if count[a%10] > 0 { return false }
        count[a%10] += 1
        a /= 10
    }

    while b > 0 {
        if count[b%10] > 0 { return false }
        count[b%10] += 1
        b /= 10
    }

    while c > 0 {
        if count[c%10] > 0 { return false }
        count[c%10] += 1
        c /= 10
    }

    for i in 0...count.count-1 {
        if count[i] != 1{
            return false
        }
    }

    return true
}

var panProduct: [Int] = [0]
var sum = 0
let limit = 2500

for a in 0...limit {
    for b in a...limit {
        let c = a * b
        if isPandigital(a, b:b, c:c) {
            var add = true
            for i in 0...panProduct.count-1 {
                if panProduct[i] == c {
                    add = false
                    break
                }
            }
            if add == true {
                panProduct.append(c)
                sum += c
            }
        }
    }
}

print(sum)

File: problem_032.go
// -*- compile-command: "go build -o problem_032_go problem_032.go" -*-
// Copyright (c) 2017 Michael Caldwell
package main

import "fmt"

func deconstruct(n int, d []int) []int {
        for n > 0 {
                d[n % 10]++
                n /= 10
        }
        return d
}

func isPandigital(a int, b int) bool {
        d := []int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0}

        d = deconstruct(a, d)
        d = deconstruct(b, d)
        d = deconstruct(a * b, d)

        if d[0] != 0 {
                return false
        }

        for i := 1; i < len(d); i++ {
                if d[i] != 1 {
                        return false
                }
        }

        return true
}

func main() {
        fmt.Println("Project Euler - Problem 32:")
        fmt.Println("Find the sum of all numbers that can be written as pandigital products.\n")

        sum := 0
        limit := 2500
        history := []int{}

        for a := 1; a < limit; a++ {
                for b := a; b < limit; b++ {
                        if isPandigital(a, b) {
                                product := a * b
                                unique := true

                                // Search for uniqueness of product
                                for i := 0; i < len(history); i++ {
                                        if history[i] == product {
                                                unique = false
                                                break
                                        }
                                }

                                // Record product for future reference
                                if unique == true {
                                        sum += product
                                        history = append(history, product)
                                }
                        }
                }
        }

        fmt.Println(sum)
}
